Multigres cluster cr design (#58)
multigres/multigres-operator: plans/phase-1/api-design/multigres-operator-api-v1alpha1-design.md

  * **Single Source of Truth:** The `MultigresCluster` is the only editable entry point for cluster topology, preventing conflicting states.
  * **Scoped Reusability:** By splitting templates into `CoreTemplate`, `CellTemplate`, and `ShardTemplate`, we provide clear, reusable configurations.
  * **Explicit Topology:** Removing "shard count" partitioning in favor of explicit shard list definitions provides deterministic control over where exactly data lives.
  * **Consistent Override Chain:** All components follow a predictable 4-level override chain, providing maximum flexibility while maintaining a clear and consistent API pattern.

## Proposal: API Architecture and Resource Topology

  * **Core Components:** `MultiAdmin` and `GlobalTopoServer` are defined as top-level fields in the `MultigresCluster`. Each can be configured inline or by referencing a `CoreTemplate`.
  * **Cells:** Explicitly defined in the root CR; can be specified inline or via a `CellTemplate`.
  * **Databases:** Follows a strict physical hierarchy: `Database` -\> `TableGroup` -\> `Shard`.
  * **Shards:** Can be specified inline or via a `ShardTemplate`.

>   * **MultiOrch:** Located at the **Shard** level to provide dedicated orchestration per Raft group.

Author: Corrected on a58fb91

```ascii
[MultigresCluster] 🚀 (Root CR - user-editable)
      │
      ├── 📍 Defines [TemplateDefaults] (Cluster-wide default templates)
      │
      ├── 🌍 [GlobalTopoServer] (Child CR) ← 📄 Uses [CoreTemplate] OR inline [spec]
      │
      ├── 🤖 MultiAdmin Resources ← 📄 Uses [CoreTemplate] OR inline [spec]
      │
      ├── 💠 [Cell] (Child CR) ← 📄 Uses [CellTemplate] OR inline [spec]
      │    │
      │    ├── 🚪 MultiGateway Resources
      │    └── 📡 [LocalTopoServer] (Child CR, optional)
      │
      └── 🗃️ [TableGroup] (Child CR)
           │
           └── 📦 [Shard] (Child CR) ← 📄 Uses [ShardTemplate] OR inline [spec]
                │
                ├── 🧠 MultiOrch Resources (Deployment/Pod)
                └── 🏊 Pools (StatefulSets for Postgres+MultiPooler)

📄 [CoreTemplate] (User-editable, scoped config)
   ├── globalTopoServer

>    └── multiadmin

Author: The problem with AdminTemplate is that it feels is only for MultiAdmin. Leaving it as it is for now until we can think of a better name.

📄 [CellTemplate] (User-editable, scoped config)
   ├── multigateway
   └── localTopoServer (optional)

📄 [ShardTemplate] (User-editable, scoped config)
   ├── multiorch
   └── pools (postgres + multipooler)
```

## Design Details: API Specification

### User Managed CR: MultigresCluster

>   * This CR and the three scoped templates (`CoreTemplate`, `CellTemplate`, `ShardTemplate`) are the *only* editable entries for the end-user.

Author: Clarified on 041e9fd

>   * All other child CRs will be owned by this top-level CR. Any manual changes to those child CRs will be automatically reverted by their respective controllers to match the `MultigresCluster` definition.

Reviewer: The rollback won't be done by the respective controllers, as those controllers will be checking for more targeted resource and provision Pods, etc. It is `MultigresCluster` controller who will revert any manual changes made to the child resources.

```yaml
apiVersion: multigres.com/v1alpha1
kind: MultigresCluster
metadata:
  name: example-cluster
  namespace: example
spec:
  # ----------------------------------------------------------------
  # Global Images Cluster Configuration
  # ----------------------------------------------------------------
  # Images are defined globally to ensure version consistency across
  # all cells and shards.
  # NOTE: Perhaps one day we can template these as well.
  images:
    imagePullPolicy: "IfNotPresent"
    imagePullSecrets:
      - name: "my-registry-secret"
    multigateway: "multigres/gateway:latest"
    multiorch: "multigres/orch:latest"
    multipooler: "multigres/pooler:latest"
    multiadmin: "multigres/admin:latest"
    postgres: "postgres:15.3"

  # ----------------------------------------------------------------
  # Cluster-Level Template Defaults
  # ----------------------------------------------------------------
  # These are the default templates to use for any component that
  # does not have an inline 'spec' or an explicit 'templateRef'.
  templateDefaults:
    coreTemplate: "cluster-wide-core"
    cellTemplate: "cluster-wide-cell"

>     shardTemplate: "cluster-wide-shard"

Author: Corrected on a58fb91

  # ----------------------------------------------------------------
  # Global Components (REVISED)
  # ----------------------------------------------------------------
  
  # Global TopoServer is a singleton. It follows the 4-level override chain.
  # It supports EITHER a 'managedSpec' OR 'external' OR 'templateRef'.
  globalTopoServer:
    # --- OPTION 1: Inline Managed Spec ---

>     managedSpec:

Reviewer: I don't think I asked this before -- why do we call it `managedSpec`, rather than just `spec`?

Author: @rytswd To separated from the external block and show that this is managed. I suppose for consistency we could call it just managed, or rename the external to externalSpec. That may make more sense

Author: Corrected on f1afd1f

      replicas: 3
      storage:
        size: "10Gi"
        class: "standard-gp3"
      resources:
        requests:
          cpu: "500m"
          memory: "1Gi"
        limits:
          cpu: "1"
          memory: "2Gi"

    # --- OPTION 2: Inline External Spec ---
    # external:
    #   endpoints: 
    #     - "https://etcd-1.infra.local:2379"
    #     - "https://etcd-2.infra.local:2379"
    #   caSecret: "etcd-ca-secret"
    #   clientCertSecret: "etcd-client-cert-secret"

    # --- OPTION 3: Explicit Template Reference ---
    # templateRef: "my-explicit-core-template"

> 

Reviewer: For completeness, I would add the option 4 as you mention 4-level override

```suggestion
    # --- OPTION 3: Explicit Template Reference ---
    # templateRef: "my-explicit-core-template"

    # --- OPTION 4: Use Cluster Template ---
    # When no spec details provided, it falls back to using templateDefaults.

```

  # MultiAdmin is a singleton. It follows the 4-level override chain.
  # It supports EITHER a 'managedSpec' OR a 'templateRef'.
  multiadmin:
    # --- OPTION 1: Inline Managed Spec ---
    managedSpec:
      replicas: 2
      resources:
        requests: 
          cpu: "200m"
          memory: "256Mi"
        limits:
          cpu: "500m"
          memory: "512Mi"

    # --- OPTION 2: Explicit Template Reference ---
    # templateRef: "my-explicit-core-template"

  # ----------------------------------------------------------------
  # Cells Configuration
  # ----------------------------------------------------------------
  cells:
    # --- CELL 1: Using an Explicit Template ---
    - name: "us-east-1a"
      # Location must be strictly one of: 'zone' OR 'region'
      zone: "us-east-1a" 
      # region: "us-east-1"
      

>       # Precedence 1: Explicit template reference

Reviewer: What does "precedence 1" mean? I see only "precedence 1" mentioned, no precedence 2 or similar?

Author: Corrected on f1e9d56

               cpu: "1"
               memory: "2Gi"
         # --- Optional Local TopoServer ---
         # If omitted, this cell uses the GlobalTopoServer.
         # localTopoServer:
         #   managedSpec:
         #      replicas: 3
         #      storage:
         #        size: "5Gi"
         #        class: "standard-gp3"
    
    # --- CELL 3: Using Cluster Default Template ---
    - name: "us-east-1c"
      zone: "us-east-1c"
      # 'spec' and 'cellTemplate' are omitted.
      # This will use 'spec.templateDefaults.cellTemplate' ("cluster-wide-cell")
      # (if that is not set, it will look for 'CellTemplate' named 'default',
      # and if that is not found, it will use the webhook default).

  # ----------------------------------------------------------------
  # Database Topology (Database -> TableGroup -> Shard)
  # ----------------------------------------------------------------
  databases:
    - name: "production_db"
      tablegroups:
        - name: "orders_tg"
          # Shards are strictly explicitly defined. 
          # No 'partitioning: { count: 10 }' auto-generation.
          shards:
            # --- SHARD 0: Using an Explicit Template ---

>             - name: "0"

Author: I don't think so. I am guessing at one point we would want to be able to define keyrange here, but the name would seem like an odd way to put it. Maybe leave it for now, but perhaps at one point naming the shards will be irrelevant, since it's a list and what matters the most is the keyrange I think.

      # MultigresCluster CR if left empty.
      # Alternatively, it can be set to a generic value if this template 
      # is specific to a region (e.g., "us-east-template").
      cell: "" 
      replicas: 2
      storage:
        class: "standard-gp3"
        size: "100Gi"
      postgres:
        requests:
          cpu: "2"
          memory: "4Gi"
        limits:
          cpu: "4"
          memory: "8Gi"
      multipooler:
        requests:
          cpu: "1"
          memory: "512Mi"
        limits:
          cpu: "2"
          memory: "1Gi"

    dr-replica:
      type: "readOnly"
      cell: "us-west-2a" # Hardcoded cell example in template
      replicas: 1
      storage:
         class: "standard-gp3"
         size: "100Gi"
      postgres:
         requests:
           cpu: "1"
           memory: "2Gi"
         limits:
           cpu: "2"
           memory: "4Gi"
      multipooler:
         requests:
           cpu: "500m"
           memory: "512Mi"
         limits:
           cpu: "1"
           memory: "1Gi"
```

### Child Resources (Read-Only)

These resources are created and reconciled by the `MultigresCluster` controller.

> > NOTE: At some point we may want to consider adding status fields for the children to say what template the config is coming from, for simplicity not defining that now. 

Author: Happy to add labels and annotation configuration options everywhere, even when not explicitly mentioned on the design doc. It makes sense

#### Child CR: TopoServer

  * Applies to both Global (owned by `MultigresCluster`) and Local (owned by `Cell`) topology servers.

>   * This CR does *not* exist if the user configures an `external` etcd connection in the parent.

Author: Corrected on a58fb91

  replicas: 3
  storage:
    size: "10Gi"
    class: "standard-gp3"
  image: "quay.io/coreos/etcd:v3.5"
  resources:
    requests:
      cpu: "500m"
      memory: "1Gi"
    limits:
      cpu: "1"
      memory: "2Gi"
status:
  conditions:
    - type: Available
      status: "True"
      lastTransitionTime: "2025-11-07T12:01:00Z"
  clientService: "example-cluster-global-client"
  peerService: "example-cluster-global-peer"
```

#### Child CR: Cell

  * Owned by `MultigresCluster`.
  * Strictly contains `MultiGateway` and optional `LocalTopoServer`. `MultiOrch` is NO LONGER here (moved to Shard).
  * The `allCells` field is used for discovery by gateways.

```yaml
apiVersion: multigres.com/v1alpha1
kind: Cell
metadata:
  name: "example-cluster-us-east-1a"
  namespace: example
  labels:
    multigres.com/cluster: "example-cluster"
    multigres.com/cell: "us-east-1a"
  ownerReferences:
    - apiVersion: multigres.com/v1alpha1
      kind: MultigresCluster
      name: "example-cluster"
      uid: "a1b2c3d4-1234-5678-90ab-f0e1d2c3b4a5"
      controller: true
spec:
  name: "us-east-1a"
  zone: "us-east-1a"

  # Images passed down from global configuration
  images:

>     multigateway: "multigres/gateway:latest"

Author: This is because in the past there used to be more than one image defined here (multiorch was here as well). 

But your comment made me realised this is missing this in shard and tablegroup, which NOT be the case. Need to fix that. 

In regards to resolving it to multiGateway.image... makes sense I suppose, but then we should do the same for shard and tablegroup as well, which would me mean quite a bit of repetition of the image key.

>     multigateway: "multigres/gateway:latest"

Reviewer: I disagree. While the consistency is important, the child CRs are there to provide "clarity" for the given component, rather than ease of management (because the expectation is that you don't manage these resources directly).

I am not saying we should change the structure of MultigresCluster resource, but update child CRs to provide a very similar representation as Deployment / StatefulSet to make it friendly for operational needs such as monitoring.

  # Always populated by the parent controller if no local server is used.
  globalTopoServer:
    rootPath: "/multigres/global"
    clientServiceName: "example-cluster-global-client"

  # Option 1: Using the Global TopoServer (Default)
  #
  topoServer: {}

  # Option 2: Inline Definition (External)
  #
  # topoServer:
  #   external:
  #     address: "etcd-us-east-1a.my-domain.com:2379"
  #     rootPath: "/multigres/us-east-1a"

  # Option 3: Managed Local
  #
  # topoServer:
  #   managedSpec:
  #     rootPath: "/multigres/us-east-1a"
  #     image: "quay.io/coreos/etcd:v3.5"
  #     replicas: 3
  #     storage:
  #       size: "5Gi"
  #       class: "standard-gp3"

  # List of all cells in the cluster for discovery.
  allCells:
  - "us-east-1a"
  - "us-east-1b"

  # Topology flags for the Cell controller to act on.
  topologyReconciliation:
    registerCell: true
    pruneTablets: true

status:
  conditions:
  - type: Available
    status: "True"
    lastTransitionTime: "2025-11-07T12:01:00Z"
  gatewayReplicas: 3
  gatewayReadyReplicas: 3
  gatewayServiceName: "example-cluster-us-east-1a-gateway"
```

#### Child CR: TableGroup

  * Owned by `MultigresCluster`.

>   * Acts as the middle-manager for Shards. It MUST contain the full resolved specification for all shards it manages, enabling it to be the single source of truth for creating its child `Shard` CRs.

Author: Yeah this is making me think a lot, thanks for bringing it up. Below is the configuration for the local provisioner. You can see there is a configuration option:

`default-db-name: postgres`

Currently there is no configuration at all for tablegroups because the code only allows you to do default DB, default TG and default shard.

But us in the operator we need to be able to explicitly define these shards and configure them. If multigres creates this and bypasses the operator completely that may be an issue for our declarative shard configuration.

Why don't we simply add a configuration flag called `default`?

It could work like this:

1. If the user does not defines any shards or DBs, the operator automatically creates the default DB/TG with a `true` value.
2. If the user defines only one DB/TG and does not define this as default, it fails.

At some point, these default DBs/TGs maybe will have additional configuration options.

```yaml
provisioner: local
provisioner-config:
    backup-repo-path: /tmp/multigres/multigres_local/data/backups
    cells:
        zone1:
            multigateway:
                grpc-port: 15170
                http-port: 15100
                log-level: info
                path: /tmp/multigres/bin/multigateway
                pg-port: 15432
            multiorch:
                grpc-port: 15370
                http-port: 15300
                log-level: info
                path: /tmp/multigres/bin/multiorch
            multipooler:
                backup-conf: /tmp/multigres/multigres_local/data/pooler_xxtwrwpc/pgbackrest.conf
                database: postgres
                grpc-port: 15270
                grpc-socket-file: /tmp/multigres/multigres_local/sockets/multipooler-zone1.sock
                http-port: 15200
                log-level: info
                path: /tmp/multigres/bin/multipooler
                pg-port: 5432
                pooler-dir: /tmp/multigres/multigres_local/data/pooler_xxtwrwpc
                service-id: xxtwrwpc
                table-group: default
            pgctld:
                grpc-port: 15470
                grpc-socket-file: /tmp/multigres/multigres_local/sockets/pgctld-zone1.sock
                log-level: info
                path: /tmp/multigres/bin/pgctld
                pg-database: postgres
                pg-port: 5432
                pg-pwfile: /tmp/multigres/multigres_local/data/pooler_xxtwrwpc/pgpassword.txt
                pg-user: postgres
                pooler-dir: /tmp/multigres/multigres_local/data/pooler_xxtwrwpc
                timeout: 30
        zone2:
            multigateway:
                grpc-port: 15171
                http-port: 15101
                log-level: info
                path: /tmp/multigres/bin/multigateway
                pg-port: 5433
            multiorch:
                grpc-port: 15371
                http-port: 15301
                log-level: info
                path: /tmp/multigres/bin/multiorch
            multipooler:
                backup-conf: /tmp/multigres/multigres_local/data/pooler_f2vwq5hh/pgbackrest.conf
                database: postgres
                grpc-port: 15271
                grpc-socket-file: /tmp/multigres/multigres_local/sockets/multipooler-zone2.sock
                http-port: 15201
                log-level: info
                path: /tmp/multigres/bin/multipooler
                pg-port: 5433
                pooler-dir: /tmp/multigres/multigres_local/data/pooler_f2vwq5hh
                service-id: f2vwq5hh
                table-group: default
            pgctld:
                grpc-port: 15471
                grpc-socket-file: /tmp/multigres/multigres_local/sockets/pgctld-zone2.sock
                log-level: info
                path: /tmp/multigres/bin/pgctld
                pg-database: postgres
                pg-port: 5433
                pg-pwfile: /tmp/multigres/multigres_local/data/pooler_f2vwq5hh/pgpassword.txt
                pg-user: postgres
                pooler-dir: /tmp/multigres/multigres_local/data/pooler_f2vwq5hh
                timeout: 30
    default-db-name: postgres
    etcd:
        data-dir: /tmp/multigres/multigres_local/data/etcd-data
        port: 2379
        version: 3.5.9
    multiadmin:
        grpc-port: 15070
        http-port: 15000
        log-level: info
        path: /tmp/multigres/bin/multiadmin
    root-working-dir: /tmp/multigres/multigres_local
    topology:
        backend: etcd2
        cells:
            - name: zone1
              root-path: /multigres/zone1
            - name: zone2
              root-path: /multigres/zone2
        global-root-path: /multigres/global
```

>   * Acts as the middle-manager for Shards. It MUST contain the full resolved specification for all shards it manages, enabling it to be the single source of truth for creating its child `Shard` CRs.

Author: Modified on 2a7902a

    
    # This shard's spec is resolved from a template
    # (e.g., "cluster-wide-shard")
    - name: "2"
      multiOrch:
        replicas: 1
        image: "multigres/orch:latest"
        resources:
          requests:
            cpu: "100m"
            memory: "128Mi"
          limits:
            cpu: "200m"
            memory: "256Mi"
      pools:
        primary:
          cell: "us-east-1c" # Resolved from override
          type: "readWrite"
          replicas: 2 # Assuming '2' from template
          storage:
             size: "100Gi" # Assuming '100Gi' from template
             class: "standard-gp3" # Assuming 'standard-gp3' from template
          postgres:
              requests: # Assuming values from template
                cpu: "2"
                memory: "4Gi"
              limits:
                cpu: "4"
                memory: "8Gi"
          multipooler:
              requests: # Assuming values from template
                cpu: "1"
                memory: "512Mi"
              limits:
                cpu: "2"
                memory: "1Gi"
        
status:
  readyShards: 3
  totalShards: 3
```

#### Child CR: Shard

  * Owned by `TableGroup`.

>   * Now contains `MultiOrch` (Raft leader helper) AND `Pools` (actual data nodes).

Author: Corrected on a58fb91

    * **Uniqueness:** Validates that all names are unique within their respective scopes:
        * Cell names in `spec.cells`.
        * Database names in `spec.databases`.
        * TableGroup names within a Database.
        * Shard names within a TableGroup.
    * **Topology Integrity:** Verifies that if a Shard is pinned to a specific cell (via overrides), that cell exists in the `spec.cells` list.

#### `CoreTemplate`, `CellTemplate`, `ShardTemplate`
* **On `CREATE` and `UPDATE`:**
    * **Schema Validation:** Ensures the template spec contains valid configuration for its specific type (e.g., a `CellTemplate` cannot contain `multiAdmin` config).

---

### 3. Asynchronous Controller and Finalizer Logic

Asynchronous logic is used for operations that depend on external state or require blocking deletion, handled by controllers and finalizers.

#### `MultigresCluster`
* **Deletion Protection (Finalizer):**
    1.  The `MultigresCluster` controller adds a finalizer (e.g., `multigres.com/cleanup`) to the CR upon creation.
    2.  On deletion, the controller ensures all child resources (StatefulSets, PVCs, Services) are properly terminated before removing the finalizer.

>     3.  *Note:* Since databases are embedded in the cluster CR, deleting the cluster implies deleting all databases.

Reviewer: We should probably mention how the deletion will delete most components, but will need to leave PVs detached and available for some time?

* **Default DB Schema (System Catalog) Availability Dependency:** The Operator's reconciliation loop now strictly depends on the read availability of the "Default Database" (System Catalog). If the Default Database is down or locked, the Operator becomes "blind" and cannot reconcile *any* part of the cluster, unlike a pure CRD model which relies only on the highly available Kubernetes API. This may not be much of an issue as either way, if the default DB is down, there it wouldn't work either way. 
* **API Hotspot and "Blast Radius" Risk:** By centralizing all database definitions into the monolithic `MultigresCluster` CR, this single object becomes a massive reconciliation hotspot. A single typo in this large resource (e.g., while "adopting" a database) could potentially break reconciliation for the entire cluster's control plane.
* **Rename/Replace Ambiguity:** Postgres allows `ALTER DATABASE RENAME`, but Kubernetes relies on stable names. If a user renames a database in SQL, the Operator may perceive this as a "Delete" (of the old name) and "Create" (of the new name), potentially attempting to re-provision the old name if it still exists in the YAML.

> * **Lack of Namespace Isolation:** All databases must be defined in (or adopted into) the central `MultigresCluster` resource. This effectively forces all application teams to rely on platform admins to manage resource sizing, removing the ability to use Kubernetes RBAC for self-service resource management in separate namespaces. No clean DBA/Platform Engineer persona separation model. 

Reviewer: Could you explain this more in detail? The namespace will be tied to the target MultigresCluster, but that's somewhat expected I thought. That's the normal multi-tenancy deployment, where each namespace has a list of resources.

Author: In the alternative DDL design you could have MultigresCluster in one namespace and the DBs on other namespaces, effectively separated DBAs/App devs from SRE/Platform Engineers. I suppose it's not really a drawback if the use case makes sense to design it without this separation, which it does, at least for now.

Defining these entries allows the user to explicitly configure the resources (replicas, storage, compute) allocated to these system components, rather than relying on hardcoded operator defaults.

### User Managed CR: MultigresCluster

  * This CR and the three scoped templates (`CoreTemplate`, `CellTemplate`, `ShardTemplate`) are the *only* editable entries for the end-user.
    * All other child CRs will be owned by this top-level CR. Any manual changes to those child CRs will be immediately reverted by the `MultigresCluster` cluster controller.
  * All component configurations (`globalTopoServer`, `multiadmin`, `cells`, `shards`) follow a consistent pattern: they can be defined via an inline `spec` or by referencing a template (`templateRef`). Providing both is a validation error.
    * **Override Chain:** All components use the following 4-level precedence chain for configuration:
      1.  **Component-Level Definition:** An inline `spec` or an explicit `templateRef` on the component itself.

>       2.  **Cluster-Level Default:** The corresponding template defined in `spec.templateDefaults` (e.g., `templateDefaults.coreTemplate` or `templateDefaults.cellTemplate`).

Reviewer: I had to refer back to this section multiple times. I think this wording is VERY confusing, because I would think "cluster" would mean the Kubernetes cluster as a whole, similar to `ClusterRole` or `ClusterRoleBinding`. I understand this is about `MultigresCluster` resource, and it should be very clear that's the case.

```suggestion
      2.  **Defaults in MultigresCluster spec:** The corresponding template defined in `spec.templateDefaults` (e.g., `templateDefaults.coreTemplate` or `templateDefaults.cellTemplate`).
```

  multiorch:
    # replicas: 1 # replicas per cell and pool this multiorch is deployed
    cells: [] # Defaults to all populated cells
    resources:
      requests:
        cpu: "100m"
        memory: "128Mi"
      limits:
        cpu: "200m"
        memory: "256Mi"

  # MAP STRUCTURE: Keyed by pool name for safe targeting.
  pools:
    primary:
      type: "readWrite"
      # 'cells' can be left empty here or omitted entirely. It MUST be overridden in the 
      # MultigresCluster CR if left empty or missing.
      # Alternatively, it can be set to a generic value if this template 
      # is specific to a region (e.g., "us-east-template").
      cells: [] 
      replicasPerCell: 2
      storage:
        class: "standard-gp3"
        size: "100Gi"
      postgres:
        requests:
          cpu: "2"
          memory: "4Gi"
        limits:
          cpu: "4"
          memory: "8Gi"
      multipooler:
        requests:
          cpu: "1"
          memory: "512Mi"
        limits:
          cpu: "2"
          memory: "1Gi"

    dr-replica:
      type: "readOnly"
      # This pool will be deployed to all cells listed here.
      cells: 
        - "us-west-2a" 

>       replicas: 1

Reviewer: ```suggestion
      replicasPerCell: 1
```

* **Pool Placement:** `pools` uses a `cells` list. For `readWrite` pools, this list typically contains only a few cells rather than using all available cells. For `readOnly` pools, this list can contain multiple cells to apply the same configuration across multiple zones and regions.

```yaml
apiVersion: multigres.com/v1alpha1
kind: ShardTemplate
metadata:
  name: "standard-shard-ha"
  namespace: example
spec:
  # Template strictly defines only Shard-scoped components.

  # MultiOrch is a shard-level component.
  # The Operator will deploy one instance of this Deployment into EVERY Cell 
  # listed in 'cells'. If 'cells' is empty, it defaults to all cells 
  # where pools are defined.
  multiorch:
    # replicas: 1 # replicas per cell and pool this multiorch is deployed
    cells: [] # Defaults to all populated cells
    resources:
      requests:
        cpu: "100m"
        memory: "128Mi"
      limits:
        cpu: "200m"
        memory: "256Mi"

  # MAP STRUCTURE: Keyed by pool name for safe targeting.
  pools:

>     primary:

Author: Don't have a very strong opinion what to name it here because it is mostly for reference, especially at this stage. Feel free to give me a commit suggestion to whatever name you like best and I'll apply it.

When the user does a `kubectl get multigrescluster minimal-cluster -o yaml` after apply this they would see all the values materialized, the default will be applied via webhook.

### 2\. The Minimalist (Relying on Cluster Defaults)

This user relies on the `spec.templateDefaults` field to set cluster-wide defaults.

```yaml
apiVersion: multigres.com/v1alpha1
kind: MultigresCluster
metadata:
  name: minimal-cluster-with-defaults
spec:
  # This cluster will use the "dev-defaults" CoreTemplate for its
  # global components. All cells and shards will use their respective
  # "dev-defaults" templates.
  templateDefaults:
    coreTemplate: "dev-defaults"
    cellTemplate: "dev-defaults-cell"
    shardTemplate: "dev-defaults-shard"

  # CoreComponents (globalTopoServer, multiadmin) are omitted,
  # so they will use "dev-defaults" (from the CoreTemplate)

  cells:
    - name: "us-east-1a"
      zone: "us-east-1a"
      # 'spec' and 'cellTemplate' are omitted, so "dev-defaults-cell" is used.
    - name: "us-west-2a"
      zone: "us-west-2a"
      # 'spec' and 'cellTemplate' are omitted, so "dev-defaults-cell" is used.

  databases:
    - name: "db1"
      tablegroups:
      - name: "tg1"

>         shards:

Reviewer: Where is the Multigres's default database + tablegroup? Are they going to be added by mutating webhook?

Reviewer: I'm not sure what you mean by the hybrid approach. Do you mean that when it is not specified, the mutating webhook will be placing the default database + tablegroup to the spec anyways? If that's what you mean, I like that approach -- I just think it's better to have explicit update to the spec, rather than controller handling the update behind the scenes.

Author: Yes, that's what I mean.

  * **2025-11-11:** Introduced a consistent 4-level override chain (inline/explicit-template -\> cluster-default -\> namespace-default -\> webhook) for all components. Added `CoreTemplate` CRD and `spec.templateDefaults` block to support this. Reverted `spec.coreComponents` nesting to top-level `globalTopoServer` and `multiadmin` fields.
  * **2025-11-14:** Explored a multi-CRD "claim" model (`MultigresDatabase` + `MultigresDatabaseResources`) to support DDL-driven workflows and strong RBAC separation.
  * **2025-11-18:** Reverted to the 2025-11-11 monolithic `MultigresCluster` design to align with client requirements for Postgres-native "System Catalog" management. The `MultigresDatabase` CRD was rejected. We will instead support the DDL workflow via a synchronization controller that patches the monolithic `MultigresCluster` CR based on the `SystemCatalog` state.
  * **2025-11-25:** Moved `multiorch` and `pool` placement from single `cell` fields to explicit `cells` lists. This supports multi-cell pool definitions (e.g., for uniform read replicas) and decouples MultiOrch placement from pool presence, while maintaining safety via a "defaults to all" logic.

## Drawbacks

> We have reverted to this design but the requirement to create resources via DDL (`CREATE DATABASE`) hasn't been dropped, but postponed. What follows are some caveats when following the current design as it stands right now and also incorporating this Posgres compatibility requirement:

Reviewer: A new reader would be confused about DDL mentioned all of a sudden. We could do a bit more clarification like this?

```suggestion
The requirement to create resources via DDL (`CREATE DATABASE`) to provide Postgres native interface was a part of the design discussions .This has not been dropped, but postponed. What follows are some caveats when following the current design as it stands right now and also incorporating this Posgres compatibility requirement:
```

>   * **Broken "Delete" UX (Zombie Resources):** To support the "DB is Truth" requirement, the Operator **cannot** delete a database simply because it is removed from the `spec.databases` list as it might still exist in the default database schema ("system catalog"). This breaks the standard Kubernetes expectation that "deleting config = deleting resource." Users must use imperative SQL (`DROP DATABASE`) to delete resources; removing them from Git will only orphan them, leaving them running and accruing costs ("Zombie Databases"). Including a bidirectional update process here may complicate things.

Reviewer: This is not quite true. We should be able to add finalizer, and ensure that finalizer check logic would trigger the external resource clean-up -- in this case, going into the default database, and drop the database from Postgres. External resource clean-up using finalizer is common, and in a way the reason why we use finalizers.

This is still an issue as operator would need to know a lot about the Postgres / Multigres internal, though, so it is still a problem we need to find a right solution. However, this isn't a problem with the deletion logic, but more about the separation of concern is difficult to achieve.

         #      storage:
         #        size: "5Gi"
         #        class: "standard-gp3"
    
    # --- CELL 3: Using Cluster Default Template ---
    - name: "us-east-1c"
      zone: "us-east-1c"
      # 'spec' and 'cellTemplate' are omitted.
      # This will use 'spec.templateDefaults.cellTemplate' ("cluster-wide-cell")
      # (if that is not set, it will look for 'CellTemplate' named 'default',
      # and if that is not found, it will use the webhook default).

  # ----------------------------------------------------------------
  # Database Topology (Database -> TableGroup -> Shard)
  # ----------------------------------------------------------------
  databases:
    # --- EXAMPLE 1: Configuring the System Default Database ---
    # This entry targets the system-level database created during bootstrap.
    # We mark it as 'default: true' to apply this configuration to the 
    # bootstrap resources (instead of creating a new user database).
    - name: "postgres"
      default: true
      tablegroups:
        - name: "default"
          default: true
          shards:
            - name: "0"
              # define resources for the system default shard

>               shardTemplate: "standard-shard-ha"

Author: At this stage I would want this document to be truthful and accurate spec guide, not so much an accurate representation of what would actually be generated if this cluster was deployed (the child specs may not match what you see on the cluster spec in this document). 

I believe when we started it was accurate, but we have changed so many things along the way that we lost some of that accuracy, and now I am feeling a little coy changing too much for fear I will introduce other typos when doing so that would affect implementation.

>               shardTemplate: "standard-shard-ha"

Reviewer: Although I believe we should have a very definitive document in place first, so that we can easily refer back to it as the design plan, I am also aware how lengthy this document has become. I'm tempted to break this into multiple files at this point, but that would be a lot of work, too, and probably not worth the time right now.

For now, at least I'd like to see some comment saying some parts are skipped for brevity. Without any clarification, it seems to indicate there is some implicit controller logic, but in reality, we are pushing for explicit control and clarity, which we should highlight more.

      controller: true
spec:
  # Resolved from MultigresCluster.
  replicas: 3
  storage:
    size: "10Gi"
    class: "standard-gp3"
  image: "quay.io/coreos/etcd:v3.5"
  resources:
    requests:
      cpu: "500m"
      memory: "1Gi"
    limits:
      cpu: "1"
      memory: "2Gi"
status:
  conditions:
    - type: Available
      status: "True"
      lastTransitionTime: "2025-11-07T12:01:00Z"
  clientService: "example-cluster-global-client"
  peerService: "example-cluster-global-peer"
```

#### Child CR: Cell

  * Owned by `MultigresCluster`.
  * Strictly contains `MultiGateway` and optional `LocalTopoServer`.
  * The `allCells` field is used for discovery by gateways.

```yaml
apiVersion: multigres.com/v1alpha1
kind: Cell
metadata:
  name: "example-cluster-us-east-1a"
  namespace: example
  labels:
    multigres.com/cluster: "example-cluster"
    multigres.com/cell: "us-east-1a"
  ownerReferences:
    - apiVersion: multigres.com/v1alpha1
      kind: MultigresCluster
      name: "example-cluster"
      uid: "a1b2c3d4-1234-5678-90ab-f0e1d2c3b4a5"
      controller: true
spec:
  name: "us-east-1a"
  zone: "us-east-1a"

  # Images passed down from global configuration
  images:

>     multigateway: "multigres/multigres:latest"

Reviewer: Regardless of whether we have image definition at top level or per `spec.multigateway`, the image details would flow down from MultigresCluster CR, which defines the images at the top level.

Because of the tight control we want to add for image versions used, I can understand the top level image list for MultigresCluster CR. However, child CRs are not meant to be manually edited, but still extremely useful for operation teams to understand the deployment setup. Especially if we were to provide canary and more complex version strategies, the top level definitions won't be sufficient. Also, other than Vitess, I personally don't know of any CNCF projects that define the image at the top level only. Those two are the main reasons I'm not comfortable with the current design.

You mention `Cell` being the only CR, but because `TableGroup` and `Shard` are structured exactly the same way, you can argue it's just `TableGroup`  (which also has `Shard`s inside) manages their images differently. This makes sense to me, because each TableGroup and Shard may not be related to each other, but as a part of MultigresCluster, there is a need of managing all of the versions together.

`Cell` CR is much less dynamic than `TableGroup` + `Shards`.

---

Considering the future version management needs, I think we should have it per resource. But at the same time, I am absolutely fine with having the top level definition similar to `MultigresCluster`, `TableGroup`, and `Shard`, **provided that** it doesn't say `images`. The fact that it says `images` tells me we are trying to generalise too much, where there is actually only a single image that can be configured at the top level.

>     multigateway: "multigres/multigres:latest"

Reviewer: I'm still struggling to make sense of how we want to have:

```yaml
spec:
  multigateway:
    replicas: 3
    resources:
      # ...
  multigatewayImage: multigres/multigres:v0.1.0
```

Rather than 

```yaml
spec:
  multigateway:
    replicas: 3
    resources:
      # ...
    image: multigres/multigres:v0.1.0
```

I thought about how the first syntax allows omitting the entire `multigateway` stanza and fall back to using template details. But our template design is to not carry over the template reference in the child CR, but make explicit spec in the child CR -- meaning all the details are provided in the child CR somewhere. It's only a matter of where to store the data.

Can you give me a reason or two why we want to take the first approach, other than the fact that TableGroup and Shard has a corresponding "images" stanza, and Cell has none?

  globalTopoServer:
    address: "example-cluster-global-client.example.svc.cluster.local:2379"
    rootPath: "/multigres/global"
    implementation: "etcd2"

  # Option 1: Using the Global TopoServer (Default)
  #
  topoServer: {}

  # Option 2: Inline Definition (External)
  #
  # topoServer:
  #   external:
  #     address: "my-etcd.some-namespace.svc.cluster.local:2379"
  #     rootPath: "/multigres/us-east-1a"
  #     implementation: "etcd2"

  # Option 3: Managed Local
  #
  # topoServer:
  #   etcd:
  #     rootPath: "/multigres/us-east-1a"
  #     image: "quay.io/coreos/etcd:v3.5"
  #     replicas: 3
  #     storage:
  #       size: "5Gi"
  #       class: "standard-gp3"

  # List of all cells in the cluster for discovery.
  allCells:
  - "us-east-1a"
  - "us-east-1b"

  # Topology flags for the Cell controller to act on.
  topologyReconciliation:
    registerCell: true
    pruneTablets: true

status:
  conditions:
  - type: Available
    status: "True"
    lastTransitionTime: "2025-11-07T12:01:00Z"
  gatewayReplicas: 3
  gatewayReadyReplicas: 3
  gatewayServiceName: "example-cluster-us-east-1a-gateway"
```

#### Child CR: TableGroup

  * Owned by `MultigresCluster`.

>   * Acts as the middle-manager for Shards. It MUST contain the full resolved specification for all shards it manages, enabling it to be the single source of truth for creating its child `Shard` CRs.

Reviewer: I think grammatically this would be more correct?

```suggestion
  * Acts as the middle-manager for Shards. It MUST contain the fully resolved specification for all shards it manages, enabling it to be the single source of truth for creating its child `Shard` CRs.
```

```yaml
apiVersion: multigres.com/v1alpha1
kind: TableGroup
metadata:
  name: "production-db-orders-tg"
  namespace: example
  labels:
    multigres.com/database: "production_db"
    multigres.com/tablegroup: "orders_tg"
  ownerReferences:
    - apiVersion: multigres.com/v1alpha1
      kind: MultigresCluster
      name: "example-cluster"

>       controller: true

Reviewer: I think most of the items shouldn't have randomised names, unless it's something we can simply discard or recreate, similar to how Deployment creates random name pods. For now, this added prefix should help segregate resources, and I don't believe we need more custom logic for the resource naming at this point.

spec:
  databaseName: "production_db"
  tableGroupName: "orders_tg"
  shardName: "0"

  # Images passed down from global configuration
  images:
    multiorch: "multigres/multigres:latest"
    multipooler: "multigres/multigres:latest"
    postgres: "postgres:15.3"

  # A reference to the GLOBAL TopoServer.
  globalTopoServer:
    address: "example-cluster-global-client.example.svc.cluster.local:2379"
    rootPath: "/multigres/global"
    implementation: "etcd2"
  
  # Fully resolved from parent TableGroup spec
  multiorch:
    cells: 
      - "us-east-1a"
    resources:
      requests:
        cpu: "100m"
        memory: "128Mi"
      limits:
        cpu: "200m"
        memory: "256Mi"

  pools:
    main-app:
      cells: 
        - "us-east-1a" 
      type: "readWrite"
      replicasPerCell: 2
      storage:
         size: "100Gi"
         class: "standard-gp3"
      postgres:
           requests:
             cpu: "2"
             memory: "4Gi"
           limits:
             cpu: "4"
             memory: "8Gi"
      multipooler:
           requests:
             cpu: "1"
             memory: "512Mi"
           limits:
             cpu: "2"
             memory: "1Gi"
status:

>   Cells: 

Author: ```suggestion
  cells: 
```

  * **2025-11-11:** Introduced a consistent 4-level override chain (inline/explicit-template -\> cluster-default -\> namespace-default -\> webhook) for all components. Added `CoreTemplate` CRD and `spec.templateDefaults` block to support this. Reverted `spec.coreComponents` nesting to top-level `globalTopoServer` and `multiadmin` fields.
  * **2025-11-14:** Explored a multi-CRD "claim" model (`MultigresDatabase` + `MultigresDatabaseResources`) to support DDL-driven workflows and strong RBAC separation.
  * **2025-11-18:** Reverted to the 2025-11-11 monolithic `MultigresCluster` design to align with client requirements for Postgres-native "System Catalog" management. The `MultigresDatabase` CRD was rejected. We will instead support the DDL workflow via a synchronization controller that patches the monolithic `MultigresCluster` CR based on the `SystemCatalog` state.
  * **2025-11-25:** Moved `multiorch` and `pool` placement from single `cell` fields to explicit `cells` lists. This supports multi-cell pool definitions (e.g., for uniform read replicas) and decouples MultiOrch placement from pool presence, while maintaining safety via a "defaults to all" logic.

## Drawbacks

> The requirement to create resources via DDL (`CREATE DATABASE`) to provide Postgres native interface was a part of the design discussions .This has not been dropped, but postponed. What follows are some caveats when following the current design as it stands right now and also incorporating this Posgres compatibility requirement:

Reviewer: ```suggestion
The requirement to create resources via DDL (`CREATE DATABASE`) to provide Postgres native interface was a part of the design discussions. This has not been dropped, but postponed. What follows are some caveats when following the current design as it stands right now and also incorporating this Posgres compatibility requirement:
```

* **MultiOrch Placement:** `multiorch` is deployed to the cells listed in `multiorch.cells`. If this list is empty or omitted, it defaults to all cells where pools are defined.
* **Pool Placement:** `pools` uses a `cells` list. For `readWrite` pools, this list typically contains only a few cells rather than using all available cells. For `readOnly` pools, this list can contain multiple cells to apply the same configuration across multiple zones and regions.

```yaml
apiVersion: multigres.com/v1alpha1
kind: ShardTemplate
metadata:
  name: "standard-shard-ha"
  namespace: example
spec:
  # Template strictly defines only Shard-scoped components.

  # MultiOrch is a shard-level component.
  # The Operator will deploy one instance of this Deployment into EVERY Cell 
  # listed in 'cells'. If 'cells' is empty, it defaults to all cells 
  # where pools are defined.
  multiorch:
    # replicas: 1 # replicas per cell and pool this multiorch is deployed
    cells: [] # Defaults to all populated cells
    resources:
      requests:
        cpu: "100m"
        memory: "128Mi"
      limits:
        cpu: "200m"
        memory: "256Mi"

  # MAP STRUCTURE: Keyed by pool name for safe targeting.

> pools:

Reviewer: Indent is off
```suggestion
  # MAP STRUCTURE: Keyed by pool name for safe targeting.
  pools:
```
